File: infra/ytdlp.go
package infra

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os/exec"
	"strconv"
	"strings"

	"YTcliV2/model"
)

// YtdlpClient wraps the yt-dlp binary.
type YtdlpClient struct {
	Binary string // default: "yt-dlp"
}

// NewYtdlpClient creates a default client.
func NewYtdlpClient() *YtdlpClient {
	return &YtdlpClient{Binary: "yt-dlp"}
}

// ytdlpResult maps a single JSON line from yt-dlp --dump-json.
type ytdlpResult struct {
	ID          string  `json:"id"`
	Title       string  `json:"title"`
	Channel     string  `json:"channel"`
	Duration    float64 `json:"duration"`
	ViewCount   int64   `json:"view_count"`
	LiveStatus  string  `json:"live_status"`
	WebpageURL  string  `json:"webpage_url"`
}

// Search runs yt-dlp ytsearch and returns parsed results.
func (c *YtdlpClient) Search(query string, limit int) ([]*model.Video, error) {
	searchStr := fmt.Sprintf("ytsearch%d:%s", limit, query)
	args := []string{
		searchStr,
		"--dump-json",
		"--no-download",
		"--flat-playlist",
		"--ignore-errors",
	}

	cmd := exec.Command(c.Binary, args...)
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		return nil, fmt.Errorf("yt-dlp search: %w\n%s", err, stderr.String())
	}

	var videos []*model.Video
	lines := strings.Split(strings.TrimSpace(stdout.String()), "\n")
	for i, line := range lines {
		if line == "" {
			continue
		}
		var raw ytdlpResult
		if err := json.Unmarshal([]byte(line), &raw); err != nil {
			continue
		}
		videos = append(videos, &model.Video{
			Index:    i + 1,
			ID:       raw.ID,
			Title:    raw.Title,
			Channel:  raw.Channel,
			Duration: secondsToTimeStr(int(raw.Duration)),
			Views:    formatCount(raw.ViewCount),
			IsLive:   raw.LiveStatus == "is_live" || raw.LiveStatus == "is_upcoming",
			URL:      raw.WebpageURL,
		})
	}
	return videos, nil
}

// StreamURL resolves the best audio stream URL for a video ID.
func (c *YtdlpClient) StreamURL(videoID string) (string, error) {
	url := "https://youtube.com/watch?v=" + videoID
	cmd := exec.Command(c.Binary,
		"-f", "bestaudio",
		"-g",
		"--no-download",
		"--ignore-errors",
		url,
	)
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		return "", fmt.Errorf("yt-dlp stream: %w\n%s", err, stderr.String())
	}

	streamURL := strings.TrimSpace(stdout.String())
	if streamURL == "" {
		return "", fmt.Errorf("yt-dlp returned empty stream URL")
	}
	return streamURL, nil
}

// secondsToTimeStr converts 1110 to "18:30".
func secondsToTimeStr(s int) string {
	h := s / 3600
	m := (s % 3600) / 60
	sec := s % 60
	if h > 0 {
		return fmt.Sprintf("%d:%02d:%02d", h, m, sec)
	}
	return fmt.Sprintf("%d:%02d", m, sec)
}

// formatCount converts 1234567 to "1.2m".
func formatCount(n int64) string {
	switch {
	case n >= 1_000_000_000:
		return fmt.Sprintf("%.1fb", float64(n)/1_000_000_000)
	case n >= 1_000_000:
		return fmt.Sprintf("%.1fm", float64(n)/1_000_000)
	case n >= 1_000:
		return fmt.Sprintf("%.1fk", float64(n)/1_000)
	default:
		return strconv.FormatInt(n, 10)
	}
}

File: src/ytdlp.rs
/// YouTube extraction via yt-dlp subprocess.

use anyhow::{Context, Result};
use serde::Deserialize;
use std::process::Command;

use crate::model::Video;

#[derive(Debug, Deserialize)]
struct RawVideo {
    id: String,
    title: String,
    #[serde(default)]
    channel: String,
    #[serde(default)]
    duration: Option<f64>,
    #[serde(default)]
    view_count: Option<i64>,
    #[serde(default)]
    live_status: Option<String>,
}

pub struct Ytdlp {
    pub binary: String,
    pub extra_args: Vec<String>,
}

impl Ytdlp {
    pub fn new() -> Self {
        Self {
            binary: "yt-dlp".into(),
            extra_args: Vec::new(),
        }
    }

    pub fn search(&self, query: &str, limit: u32) -> Result<Vec<Video>> {
        let search = format!("ytsearch{}:{}", limit, query);

        let mut cmd = Command::new(&self.binary);
        cmd.args([
            &search,
            "--dump-json",
            "--flat-playlist",
            "--no-download",
            "--ignore-errors",
            "--no-warnings",
        ]);

        for arg in &self.extra_args {
            cmd.arg(arg);
        }

        let output = cmd
            .output()
            .with_context(|| format!("gagal menjalankan {}", self.binary))?;

        if !output.status.success() && output.stdout.is_empty() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            anyhow::bail!("yt-dlp error: {}", stderr.trim());
        }

        let stdout = String::from_utf8_lossy(&output.stdout);
        let mut videos = Vec::new();

        for line in stdout.lines() {
            if line.is_empty() {
                continue;
            }
            match serde_json::from_str::<RawVideo>(line) {
                Ok(raw) => {
                    let dur = raw.duration.unwrap_or(0.0);
                    let id = raw.id;
                    let title = raw.title;
                    let channel = raw.channel;
                    let views = raw.view_count.map(|v| v.max(0) as u64).unwrap_or(0);
                    let is_live = matches!(
                        raw.live_status.as_deref(),
                        Some("is_live" | "is_upcoming")
                    );
                    videos.push(Video {
                        id,
                        title,
                        channel,
                        duration: dur,
                        views,
                        is_live,
                    });
                }
                Err(e) => {
                    eprintln!("warn: skip line: {}", e);
                }
            }
        }

        Ok(videos)
    }

    pub fn stream_url(&self, video_id: &str) -> Result<String> {
        let url = format!("https://youtube.com/watch?v={}", video_id);

        let mut cmd = Command::new(&self.binary);
        cmd.args([
            "-f", "bestaudio",
            "-g",
            "--no-download",
            "--ignore-errors",
            "--no-warnings",
            &url,
        ]);

        for arg in &self.extra_args {
            cmd.arg(arg);
        }

        let output = cmd
            .output()
            .with_context(|| format!("gagal menjalankan {}", self.binary))?;

        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            anyhow::bail!("yt-dlp stream error: {}", stderr.trim());
        }

        let stream_url = String::from_utf8_lossy(&output.stdout).trim().to_string();
        if stream_url.is_empty() {
            anyhow::bail!("yt-dlp mengembalikan URL kosong untuk video {}", video_id);
        }

        Ok(stream_url)
    }

    pub fn check_available() -> Result<String> {
        let output = Command::new("yt-dlp")
            .arg("--version")
            .output()
            .context("yt-dlp tidak ditemukan. Install: https://github.com/yt-dlp/yt-dlp")?;

        let version = String::from_utf8_lossy(&output.stdout).trim().to_string();
        Ok(version)
    }
}
